bash: use default PATH in FHS environments

danielfullmer opened the pull request:
If bash is executed within an environment where `PATH` is not set, it uses
the `DEFAULT_PATH_VALUE` compiled into bash to set `PATH`. In nixpkgs we set
this to `/no-such-path` by default. This makes sense in a nixpkgs/NixOS
environment since paths like `/bin` or `/usr/bin` should not be used.
However, when bash is used inside an FHS environment, this produces
results that differ from distributions which follow the FHS standard.

Before this change:
$ steam-run env -i /bin/bash -c 'echo $PATH'
/no-such-path

After this change:
$ steam-run env -i /bin/bash -c 'echo $PATH'
/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:.

---

I encountered an issue caused by this when trying to build Android 12 inside a `fhs-userenv` environment.  Somewhere deep in the android build system it clears the environment and depends on bash using its normal `DEFAULT_PATH_VALUE` to reset the `PATH` to something usable.

If `forFHSEnv` is not set, it should not change the bash derivations, and so should not require a mass-rebuild.

CC @peti @yegortimoshenko

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

r-rmcgibbo commented:
Result of `nixpkgs-review pr 126448` at 82fe6ba3 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>ovftool</li>
  </ul>
</details>
<details>
  <summary>77 packages skipped due to time constraints:</summary>
  <ul>
    <li>Sylk</li>
    <li>altair</li>
    <li>android-studio (androidStudioPackages.stable)</li>
    <li>androidStudioPackages.beta</li>
    <li>androidStudioPackages.canary</li>
    <li>androidStudioPackages.dev</li>
    <li>anki-bin</li>
    <li>appimage-run</li>
    <li>apple-music-electron</li>
    <li>arduino-ci</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>11 packages built successfully:</summary>
  <ul>
    <li>dropbox</li>
    <li>dropbox-cli</li>
    <li>flutter (flutterPackages.stable)</li>
    <li>kodiPackages.steam-launcher</li>
    <li>mate.caja-dropbox</li>
    <li>protontricks</li>
    <li>steam</li>
    <li>steam-run</li>
    <li>steam-run-native</li>
    <li>steam-tui</li>
    <li>steamcmd</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/add-bin-bash-to-avoid-unnecessary-pain/5673/53

mschwaig reviewed (approved):
This change makes sense to me for FHS environments. It does not look like it could affect anything else and I think the small number of changed packages `nixpkgs-review` picks up also confirms that.

SuperSandro2000 commented:
@jonringer can you take a look at this?

jonringer commented:
I wonder if we should expose this as something like `bashInteractiveFHS` or something in all-packages.nix

danielfullmer committed:
bash: use default PATH in FHS environments

If bash is executed within an environment where PATH is not set, it uses
the DEFAULT_PATH_VALUE compiled into bash to set PATH. In nixpkgs we set
this to /no-such-path by default. This makes sense in a nixpkgs/NixOS
environment since paths like /bin or /usr/bin should not be used.
However, when bash is used inside an FHS environment, this produces
results that differ from distributions which follow the FHS standard.

Before this change:
$ steam-run env -i /bin/bash -c 'echo $PATH'
/no-such-path

After this change:
$ steam-run env -i /bin/bash -c 'echo $PATH'
/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:.

danielfullmer commented:
@jonringer  I fixed the merge conflicts and took your suggestion to make a `bashInteractiveFHS`.

jonringer reviewed (approved):
was able to launch steam
